### PR TITLE
Document tree API for C 1.0

### PR DESCRIPTION
**Pull request by jeromekelleher**
Part of #2140

### CODE REVIEW

**Comment by benjeffery** [comment]
This only covers #2140 (tree) right? If so I'll start on #2139 (treeseq).

**Comment by jeromekelleher** [comment]
> This only covers #2140 (tree) right? If so I'll start on #2139 (treeseq).

Yep, that'd be great!

This one turned into quite a big diff (bunch of subtle things showed up when I started getting into it) so maybe we tackle the treeseq one by a few smaller updates?

**Comment by jeromekelleher** [comment]
Working my way through this one I'm wondering if TSK_NO_SAMPLE counts is worth documenting for 1.0. It adds a lot of caveats and it's quite tricky explaining what root tracking is, etc. I'm not sure it's actually used much, so maybe something to look at later to see what value we're getting from it?

What do you think @benjeffery?

**Comment by benjeffery** [comment]
Sounds good to me. Are you suggesting that it might be best to deprecate at some point?

Let me know when you'd like a review here.

**Comment by jeromekelleher** [comment]
OK, I'll pull it out for this PR and we can choose to add it back in to the documented API later if we wish. 

I haven't thought through whether we'd want to deprecate it - let's not worry about that for the moment.

**Comment by jeromekelleher** [comment]
This is ready for review @benjeffery, I'll open more specific issues to track the stuff left over.

File: c/examples/tree_iteration.c [hunk 1]
```diff
@@ -25,20 +25,20 @@ main(int argc, char **argv)
     check_tsk_error(ret);
 
     printf("Iterate forwards\n");
-    for (iter = tsk_tree_first(&tree); iter == 1; iter = tsk_tree_next(&tree)) {
+    for (ret = tsk_tree_first(&tree); ret == TSK_TREE_OK; ret = tsk_tree_next(&tree)) {
```
Comment (benjeffery) [inline c/examples/tree_iteration.c:28]: I like this change, but there's a few places it is missing:
```
test_stats.c:634
test_trees.c:303,321,352,367,6481
halotype_matching.c:982,1345
stats.c:195,232
trees.c:5448,5468
```
I also wonder if `tsk_diff_iter_next` should use the same convention?

File: c/tests/test_trees.c [hunk 2]
```diff
@@ -5382,6 +5383,43 @@ test_virtual_root_properties(void)
...
+    tsk_id_t nodes;
+    tsk_size_t n;
+
+    tsk_treeseq_from_text(&ts, 1, single_tree_ex_nodes, single_tree_ex_edges, NULL, NULL,
+        NULL, NULL, NULL, 0);
+
+    ret = tsk_tree_init(&t, &ts, TSK_NO_SAMPLE_COUNTS);
+    CU_ASSERT_EQUAL_FATAL(ret, 0);
+    ret = tsk_tree_first(&t);
+    CU_ASSERT_EQUAL_FATAL(ret, 1);
```
Comment (benjeffery) [inline c/tests/test_trees.c:5401]: TSK_TREE_OK

File: c/tskit/trees.h [hunk 3]
```diff
@@ -162,21 +165,51 @@ typedef struct {
...
+    @rst
+
+    Example:
+
+    .. code-block:: c
+
+        tsk_tree_t tree;
+        // initialise etc
+        tsk_tree_first(&tree);
+        printf("Coordinates covered by first tree are left=%f, right=%f\n",
```
Comment (benjeffery) [inline c/tskit/trees.h:182]: Maybe add a line here to check the return of `tsk_tree_first`?

File: c/tskit/trees.h [hunk 4]
```diff
@@ -396,25 +458,263 @@ int tsk_treeseq_f4(const tsk_treeseq_t *self, tsk_size_t num_sample_sets,
...
+sequence will transform it into the
+:ref:`null state<sec_c_api_trees_null>` (equivalent to
+calling :c:func:`tsk_tree_clear`).
+
+Please see the :ref:`sec_c_api_examples_tree_iteration` examples for
+recommended usage.
+@endrst
+
+@param self A pointer to an initialised tsk_tree_t object.
+@return Return 0 or 1 on success or a negative value on failure.
```
Comment (benjeffery) [inline c/tskit/trees.h:621]:
```suggestion
@return Return TSK_TREE_OK on successfully transforming to a
non-null tree; 0 on successfully transforming into the null
tree; or a negative value if an error occurs.
```

File: docs/c-api.rst [hunk 5]
```diff
@@ -382,14 +384,159 @@ Tree sequences
...
+    out:
+        tsk_tree_free(&tree);
+        return ret;
+    }
+
+
+In this example we first initialise a :c:struct:`tsk_tree_t` object,
+associating it with the input tree sequence. We then iterate over the
+trees along the sequence using a ``for`` loop, with the ``ret`` variable
+controling iteration. The usage of ``ret`` here follows a slightly
```
Comment (benjeffery) [inline docs/c-api.rst:462]:
```suggestion
controlling iteration. The usage of ``ret`` here follows a slightly
```

File: docs/c-api.rst [hunk 6]
```diff
@@ -382,14 +384,159 @@ Tree sequences
...
+different to other functions in the tskit C API
```
Comment (benjeffery) [inline docs/c-api.rst:463]:
```suggestion
different usage to other functions in the tskit C API
```

**Review by benjeffery — APPROVED** [review]
Great stuff, especially the bit explaining the tree iteration. Just a few small comments, which I can handle if you agree with them.

**Comment by jeromekelleher** [comment]
Suggestions LGTM @benjeffery, go ahead with updates if you like.

**Pull request merged**